Fix issues with None platform

hardys opened the pull request:
Since #719 None platform is broken, this aims to restore the functionality added in #630

WIP pending further testing

ironcladlou commented on hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:
```diff
@@ -2082,20 +2082,33 @@ func reconcileCAPICluster(cluster *capiv1.Cluster, hcluster *hyperv1.HostedClust
+				Namespace:  hcp.Namespace,
+				Name:       hcp.Name,
+			},
+		}
```
@enxebre @csrwng I wonder if this implies another `Platform` interface method to enable platform implementations to build the CAPI Cluster like we do with the CAPI Infrastructure (`ReconcileCAPIInfraCR()`)?

https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/hostedcluster/internal/platform/platform.go#L25

hardys replied:
Another fix for this is to make `ReconcileCAPIInfraCR` [return an object instead of nil](https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/hostedcluster/internal/platform/none/none.go#L18), so the `InfrastructureRef` can be populated, but I'm not sure if that's a valid approach (since in the None case, we're not creating any Infrastructure, we'd end up returning and presumably not persisting a generic `Cluster` object?)

ironcladlou replied:
Related prior discussion along these lines: https://github.com/openshift/hypershift/pull/719#discussion_r758638808

If some platforms don't even imply a CAPI Infra or Cluster resource at all, that to me seems like further justification for a contract more like:

```go
type Platform interface {
  DesiredCAPIInfrastructure(...) (client.Object, error)
  DesiredCAPICluster(capiInfrastructure client.Object...) (client.Object, error)
}
```

Where the call side (hostedcluster controller) handles persistence (if the impl returns anything to persist), and where the implementation may return `nil` for either if they don't require those resources.

enxebre replied:
If there's no infraCR it means there's no CAPI support. So please just return early at the top of the func:

```
if infraCR != nil {
 return nil
}
```

ironcladlou replied:
> If there's no infraCR it means there's no CAPI support. So please just return early at the top of the func:
> 
> ```
> if infraCR != nil {
>  return nil
> }
> ```
> 
> Even better only run this chunk
> 
> ```
> 	// Reconcile the CAPI Cluster resource
> 	capiCluster := controlplaneoperator.CAPICluster(controlPlaneNamespace.Name, hcluster.Spec.InfraID)
> 	_, err = createOrUpdate(ctx, r.Client, capiCluster, func() error {
> 		return reconcileCAPICluster(capiCluster, hcluster, hcp, infraCR)
> 	})
> ```
> 
> `if infra != nil`
> 
> No need to create the cluster CR at all.

This makes me wonder about the `Platform.ReconcileCAPIInfraCR` interface method. Here are its docs:

```go
// ReconcileCAPIInfraCR is called during HostedCluster reconciliation prior to reconciling the CAPI Cluster CR.
// Implementations should use the given input and client to create and update the desired state of the
// platform infrastructure CAPI CR, which will then be referenced by the CAPI Cluster CR.
// TODO (alberto): Pass createOrUpdate construct instead of client.
ReconcileCAPIInfraCR(ctx context.Context, c client.Client, createOrUpdate upsert.CreateOrUpdateFN,
```

Sounds like we're saying if the platform doesn't support CAPI then `ReconcileCAPIInfraCR` should return `nil`. If that's the case, we should document it (and document the follow on implications — does it mean this controller won't call `CAPIProviderDeploymentSpec`?). Seems like there should be a way to make this more clear. For example, what if CAPI was split out into a different interface, like:

```go
type CAPIPlatform interface {
  ReconcileCAPIInfraCR()
  CAPIProviderDeploymentSpec()
}
```

and in the controller, only call those methods at the appropriate time if the platform implements CAPI support. Then `ReconcileCAPIInfraCR()` returning `nil` would be considered an error.

Thoughts?

enxebre replied:
Yeh that all makes total sense to me as a consolidation step. This is intentionally loose atm so we can identify unknowns, platform peculiarities and better boundaries while introducing https://github.com/openshift/hypershift/pull/760 and having empiric feedback.
I'd suggest we iterate to tie the contract farther and go more rigid with something like what you are suggesting after merging this PR and https://github.com/openshift/hypershift/pull/760.

enxebre commented on support/globalconfig/infrastructure.go:
```diff
@@ -57,5 +57,8 @@ func ReconcileInfrastructure(infra *configv1.Infrastructure, hcp *hyperv1.Hosted
 		infra.Status.PlatformStatus = &configv1.PlatformStatus{}
 		infra.Status.PlatformStatus.Type = configv1.IBMCloudPlatformType
+	case hyperv1.NonePlatform:
+		infra.Status.PlatformStatus = &configv1.PlatformStatus{}
```
Please just set
`infra.Status.PlatformStatus.Type = configv1.PlatformType(hcp.Spec.Platform.Type)`
for any every case, no need to discriminate by platform.

hardys replied:
Ack thanks, I see that's already the default so I'll remove setting the Type

enxebre replied:
we can drop ibm/none specifics and just do:

```
infra.Status.PlatformStatus = &configv1.PlatformStatus{}
infra.Status.Platform = configv1.PlatformType(hcp.Spec.Platform.Type)

switch hcp.Spec.Platform.Type {
	case hyperv1.AWSPlatform:
		infra.Spec.PlatformSpec.AWS = &configv1.AWSPlatformSpec{}
		infra.Status.PlatformStatus.Type = configv1.AWSPlatformType
		infra.Status.PlatformStatus.AWS = &configv1.AWSPlatformStatus{
			Region: hcp.Spec.Platform.AWS.Region,
		}
		tags := []configv1.AWSResourceTag{}
		for _, tag := range hcp.Spec.Platform.AWS.ResourceTags {
			tags = append(tags, configv1.AWSResourceTag{
				Key:   tag.Key,
				Value: tag.Value,
			})
		}
		infra.Status.PlatformStatus.AWS.ResourceTags = tags
```

I'll do that in a follow up PR.

hardys commented:
Ok I addressed the feedback, thanks for the reviews!

Leaving as WIP since this still doesn't result in a fully working None HostedCluster in my environment - feel free to remove the hold if you want to land this part of the fix and we can work on the remainder as a follow-up.

enxebre commented:
/lgtm

enxebre commented:
/approve

>Leaving as WIP since this still doesn't result in a fully working None HostedCluster in my environment - feel free to remove the hold if you want to land this part of the fix and we can work on the remainder as a follow-up.

Thanks @hardys Feel free to unhold/continue in this PR and proceed as you feel more confortable with.

rmohr commented:
@hardys can the wip label be removed?

enxebre commented:
/hold cancel

hardys commented:
> @hardys can the wip label be removed?

Yes, although as mentioned this doesn't result in a fully working None platform deployment yet (in my test environment at least), so there may be follow-up additional fixes (help welcome finding those, I've not had time to fully debug yet)

rmohr commented:
> Yes, although as mentioned this doesn't result in a fully working None platform deployment yet (in my test environment at least), so there may be follow-up additional fixes (help welcome finding those, I've not had time to fully debug yet)

Have it. That should fix it:

```diff
diff --git a/control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go b/control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go
index 631056f7..15e46524 100644
--- a/control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go
+++ b/control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go
@@ -502,17 +502,22 @@ func (r *HostedControlPlaneReconciler) update(ctx context.Context, hostedControl
                }
        }
 
-       // If the cluster is marked paused, don't do any reconciliation work at all.
-       if cluster, err := util.GetOwnerCluster(ctx, r.Client, hostedControlPlane.ObjectMeta); err != nil {
-               return fmt.Errorf("failed to get owner cluster: %w", err)
-       } else {
-               if cluster == nil {
-                       r.Log.Info("Cluster Controller has not yet set OwnerRef")
-                       return nil
-               }
-               if annotations.IsPaused(cluster, hostedControlPlane) {
-                       r.Log.Info("HostedControlPlane or linked Cluster is marked as paused. Won't reconcile")
-                       return nil
+       switch hostedControlPlane.Spec.Platform.Type {
+       case hyperv1.NonePlatform:
+               // No CAPI cluster exists, checking for its existence is hopeless.
+       default:
+               // If the cluster is marked paused, don't do any reconciliation work at all.
+               if cluster, err := util.GetOwnerCluster(ctx, r.Client, hostedControlPlane.ObjectMeta); err != nil {
+                       return fmt.Errorf("failed to get owner cluster: %w", err)
+               } else {
+                       if cluster == nil {
+                               r.Log.Info("Cluster Controller has not yet set OwnerRef")
+                               return nil
+                       }
+                       if annotations.IsPaused(cluster, hostedControlPlane) {
+                               r.Log.Info("HostedControlPlane or linked Cluster is marked as paused. Won't reconcile")
+                               return nil
+                       }
                }
        }

```

The control plane comes up again.
